samples/net/cellular_modem/src/at_shell.c

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2024 Trackunit Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/shell/shell.h>
+#include <zephyr/modem/pipelink.h>
+
+MODEM_PIPELINK_DT_DECLARE(DT_ALIAS(modem), shell);
+
+static struct modem_pipelink *shell_pipelink = MODEM_PIPELINK_DT_GET(DT_ALIAS(modem), shell);
+static struct k_work receive_work;
+
+static uint8_t transmit_buf[255];
+static uint8_t receive_buf[255];
+
+static void receive_work_handler(struct k_work *work)
+{
+	int ret;
+	struct modem_pipe *pipe;
+
+	pipe = modem_pipelink_get_pipe(shell_pipelink);
+	if (pipe == NULL) {
+		printk("Nearly exploded, why is pipe NULL?");
+		return;
+	}
+
+	ret = modem_pipe_receive(pipe, receive_buf, sizeof(receive_buf));
+	if (ret < 1 || ret == sizeof(receive_buf)) {
+		return;
+	}
+
+	printk("%s", receive_buf);
+}
+
+static void at_shell_pipe_callback(struct modem_pipe *pipe,
+				   enum modem_pipe_event event,
+				   void *user_data)
+{
+	switch (event) {
+	case MODEM_PIPE_EVENT_OPENED:
+		printk("AT shell opened\n");
+		break;
+
+	case MODEM_PIPE_EVENT_RECEIVE_READY:
+		k_work_submit(&receive_work);
+		break;
+
+	case MODEM_PIPE_EVENT_CLOSED:
+		printk("AT shell closed\n");
+		break;
+
+	default:
+		break;
+	}
+}
+
+void pipelink_shell_callback(struct modem_pipelink *link, enum modem_pipelink_event event,
+			     void *user_data)
+{
+	ARG_UNUSED(user_data);
+
+	switch (event) {
+	case MODEM_PIPELINK_EVENT_CONNECTED:
+		modem_pipe_attach(modem_pipelink_get_pipe(link), at_shell_pipe_callback, NULL);
+		modem_pipe_open_async(modem_pipelink_get_pipe(link));
+		printk("AT shell connected\n");
+		break;
+
+	case MODEM_PIPELINK_EVENT_DISCONNECTED:
```

Comment:
What happens to the pipe when this is received, who closes it?

Reply:
The pipe is forcefully closed by the provider of the pipe. In the case of CMUX, the disconnected event would precede the CMUX being disconnected.